### DISCUSSION TITLE
How to get left and right shift to be valid responses?

### DISCUSSION THREAD

**Discussion by rpan001** [jspsych/jsPsych#1778] [Q&A]
I was looking into jspsych.js and saw that there was only one keycode value for shift, 16, which covered both the left and right shift and I was wondering if there was any way possible to have jspsych distinguish between the two.

I found this relevant stackoverflow post, but I'm unsure where in the code or how to implement it. Any help is appreciated.
[https://stackoverflow.com/questions/22029033/can-javascript-tell-the-difference-between-left-and-right-shift-key]

**Comment by becky-gilbert** [comment]
Hi @rpan001, yes, jsPsych 6.3.0+ uses the keyboard event's `.key` property, which doesn't distinguish between right and left shift (it has the value "Shift" for both, [see here for a demo](https://keycode.info/)). jsPsych versions prior to 6.3.0 used the deprecated `.keyCode` property, which would have the value of 16 for both. 

You could get jsPsych to distinguish between left and right shift, but it might be a little tricky. One option is to modify the `jsPsych.pluginAPI.getKeyboardResponse` function in the jspsych.js file ([here in the most recent release](https://github.com/jspsych/jsPsych/blob/fb2ad412242e60cad5ea9531546158aa70ab9e7b/jspsych.js#L2256-L2351), jsPsych v6.3.1). You could switch out the use of `e.key` to `e.code`, which will give you "ShiftLeft" and "ShiftRight". Keep in mind that this change would affect _all_ trials in your experiment, including the values that you're using for the key choice parameters. And it might not work without some additional modifications, but you could try it and see. 

Another option is to set up your own `keydown` event listener in the plugin file that you're using for the trials that capture right/left shift key presses, instead of using the `jsPsych.pluginAPI.getKeyboardResponse` convenience function for this. This would allow you to access the keyboard event directly, and therefore store the event's `.code` instead `.key` property. Something like this:
```js
var trial_data = {
  rt: null,
  key: null
};
var start_time = performance.now();
document.addEventListener('keydown', function(e) {
  // check if the response is valid
  if (trial.choices == jsPsych.ALL_KEYS || trial.choices.includes(e.code)) {
    // if so, then store the key press and RT
    trial_data.rt = performance.now() - start_time;
    trial_data.key = e.code;
    // might want to do other stuff here, like call the end_trial function if necessary
  }
});
```
I hope this helps. Feel free to follow up if you need further guidance.
